### ISSUE TITLE
False negative when assigning class constants inside __init__

### ISSUE THREAD

**Issue by jfmend** [microsoft/pyright#8013]
## Environment data
Tested on:
- pylance v2024.5.1 
- python 3.10.12

## Code Snippet

```python
from typing import Final

class C:
    classConstant_A: Final = 0
    classConstant_B: Final
    def __init__(self) -> None:
        C.classConstant_A += 1
        C.classConstant_B = C.classConstant_A * 10

C()
print(C.classConstant_A)  # -> 1
print(C.classConstant_B)  # -> 10
C()
print(C.classConstant_A)  # -> 2
print(C.classConstant_B)  # -> 20
```

## Expected behavior
Errors on:
- reassigning `classConstant_A` in `__init__`
- reassigning `classConstant_B` in `__init__`
- declaration of `classConstant_B` with no initialization

## Actual behavior
No errors.
This would be the correct behavior if `__init__` was initializing instance constants using `self` to access the fields instead of the class identifier `C`. However, in this example `__init__` is reassigning class variables which were declared constant with `Final`, so this should be an error.

**Comment by jfmend** [comment]
I agree with the sloppiness of the language on this respect, this issue is an easy source of confusion can easily cause bugs that are hard to identify. That is why I think a tool like PyRight should help if possible.

The spec is talking about instance variables where it says:
> A final attribute declared in a class body without an initializer must be initialized in the __init__ method (except in stub files):
```
class ImmutablePoint:
    x: Final[int]
    y: Final[int]  # Error: final attribute without an initializer

    def __init__(self) -> None:
        self.x = 1  # Good
```
This makes sense, a final attribute declared in a class body without an initializer must be intended by the user as either:

1. A final instance variable - requiring a unique assignment inside `__init__` using `self`
2. A final class variable - requiring a unique assignment at the declaration site

In my example, neither `classConstant_A` nor `classConstant_B` fulfill these requirements. So the user must have made a mistake and the mistake is plain to see:

- Variables that are declared `Final` (class variables in this case) are being modified in the program

I think the static analysis tool should point out this contradiction.

**Comment by erictraut** [comment]
This is included in pyright 1.1.365.
